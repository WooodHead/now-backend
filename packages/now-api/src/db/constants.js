/* eslint-disable import/prefer-default-export  */

export const TABLES = {
  ACTIVITY: 'now_template',
  DEVICE: 'now_device',
  EVENT: 'now_event',
  RSVP: 'now_rsvp',
  MESSAGE: 'now_messages',
  USER: 'now_user',
};
