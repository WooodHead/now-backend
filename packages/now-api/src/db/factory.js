/* eslint-disable import/no-extraneous-dependencies  */
import { Factory } from 'rosie';
import uuid from 'uuid/v4';
import { address, name, date, internet, random, lorem, hacker } from 'faker';
import randomEmoji from 'random-emoji';
import { LocalDate, ZoneId, LocalDateTime } from 'js-joda';
import {
  EVENT_INVITE_TYPE,
  APP_INVITE_TYPE,
} from '../schema/resolvers/Invitation';

const md5 = () =>
  random
    .hexaDecimal(32)
    .toLowerCase()
    .substring(2);

Factory.define('user')
  .attr('id', uuid)
  .attr('auth0Id', () => random.alphaNumeric(24))
  .attr('email', internet.email)
  .attr('firstName', name.firstName)
  .attr('lastName', name.lastName)
  .attr('bio', lorem.paragraph)
  .attr('location', () => `${address.city()}, ${address.stateAbbr()}`)
  .attr('preferences', {})
  .attr('birthday', () =>
    LocalDate.now()
      .minusYears(20)
      .toString()
  )
  .attr('photoId', md5)
  .attr(
    'photoPreview',
    'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACgAAAAoCAYAAACM/rhtAAAAAXNSR0IArs4c6QAAAmRJREFUWAntmN1Kw0AQhU/8Fwv1F0FEKCgV9UbvvKkPIFJ8AB/C1/BZfAdF9EpBFEUEBQUp0l6opRVEi/Y4LIlx006aWHuRgTHpZHbmy9lssuh81g0dbF0dzPaNlgBGnaFEwUTBqAqY8e/vwM4OUCiYiO7Ytmfw+Bi4vRXXoUlW2wD396VhqRQGD2gL4N0dcH/fwYBGPSIWiwKq/atSsFbTlvudV60CJyduPOwU97hD7WcfH8D2NpBOA0tLwOIikM0CfX32fH/06AjgCjb29ga8vEg9E2t0bArYU8/gfod3vrcnztjcnMASeGoquMXBwe9rrMUb1lhTQBYZGwMeH91yVPXqSnx3FxgZEVgqPD8PDA5K7uWl3Jg7Us4IODvrj9p/twToL/X0BBweind3A5mMgJ6e+jPld5iFogIcH7c3skW5oG5uxG3XGQuzUFSrOAxgEJQ3Hjsgn8E4LcwU/4uCr68A348aUwHGrSDBtCqqAFMpYGBAc7/6HO1zqAJk27hVjB0waCVPTAC5HDA6qlePmdopVr0HWTBIwa0t+TYz5+EBODsTN9srxm2mVVANaFNwYcGFI8T0tPj6OvD8LKDn58D19c8NA3P/HNBxgM1NtrLb8DCwtibOHQy/y1T34gKoVIByWaB7e+3jTVStoH+KV1aAmRlTpvGxvx9gPp07I34KqWAzOFZVA3qnuKu+9vP5xlBBV6k8t2p0jalfM3wPDg1JydVVYHJSUz56jhqQragip2VjI3pjbQX1FLMgd8/Ly7JB1TaImuck/8CMKGGoZzBir5aGJ4AtyeYZlCjoEaOl0y+Le57bWRVobwAAAABJRU5ErkJggg=='
  );

Factory.define('device')
  .option('user', () => ({ id: uuid() }))
  .attr('token', uuid)
  .attr('type', () => random.arrayElement(['ios', 'android']))
  .attr('model', random.randomWord)
  .attr('userId', ['user'], ({ id }) => id);

Factory.define('rsvp')
  .option('user', () => ({ id: uuid() }))
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('action', 'add')
  .attr('userId', ['user'], ({ id }) => id)
  .attr('eventId', ['event'], ({ id }) => id);

Factory.define('appInvite')
  .option('inviter', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('type', APP_INVITE_TYPE)
  .attr('code', () => String(random.number({ max: 999999, min: 100000 })))
  .attr('expiresAt', () =>
    LocalDateTime.now()
      .plusDays(1)
      .toString()
  )
  .attr('inviterId', ['inviter'], ({ id }) => id);

Factory.define('eventInvite')
  .option('inviter', () => ({ id: uuid() }))
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('type', EVENT_INVITE_TYPE)
  .attr('code', () => String(random.number({ max: 999999, min: 100000 })))
  .attr('expiresAt', () =>
    LocalDateTime.now()
      .plusDays(1)
      .toString()
  )
  .attr('inviterId', ['inviter'], ({ id }) => id)
  .attr('eventId', ['event'], ({ id }) => id);

Factory.define('activity')
  .attr('id', uuid)
  .attr('activityDate', () =>
    date
      .soon(1)
      .toISOString()
      .substring(0, 10)
  )
  .attr('title', () => `${hacker.ingverb()} for ${hacker.noun()}`)
  .attr('emoji', () => randomEmoji.random({ count: 1 })[0].character)
  .attr('description', hacker.phrase);

Factory.define('todayActivity')
  .extend('activity')
  .attr('activityDate', () => new Date().toISOString().substring(0, 10));

Factory.define('event')
  .option('activity', () => ({
    id: uuid(),
    activityDate: date
      .soon(1)
      .toISOString()
      .substring(0, 10),
  }))
  .option('location', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('activityId', ['activity'], ({ id }) => id)
  .attr('locationId', ['location'], ({ id }) => id)
  .attr('limit', () => random.number({ max: 20, min: 5 }))
  .attr('time', ['activity'], ({ activityDate: activityDateString }) => {
    const activityDate = LocalDate.parse(activityDateString);
    return date
      .between(
        activityDate.atStartOfDay(),
        activityDate.atTime4(23, 59, 59, 999)
      )
      .toISOString();
  })
  .attr('timezone', () => random.arrayElement(ZoneId.getAvailableZoneIds()));

Factory.define('location')
  .attr('id', uuid)
  .attr('name', hacker.ingverb)
  .attr('address', address.streetAddress)
  .attr('city', address.city)
  .attr('state', address.stateAbbr)
  .attr('postalCode', address.zipCode)
  .attr('country', address.countryCode)
  .attr('neighborhood', hacker.noun)
  .attr('foursquareVenueId', md5)
  .attr(
    'location',
    () => `SRID=4326;POINT(${address.longitude()} ${address.latitude()})`
  );

Factory.define('message')
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('ts', () => date.recent(1).getTime())
  .attr('text', hacker.phrase)
  .attr('eventId', ['event'], ({ id }) => id)
  .attr('userId', ['user'], ({ id }) => id);

export default Factory;
