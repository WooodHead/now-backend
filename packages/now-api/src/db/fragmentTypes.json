{
  "__schema": {
    "types": [
      {
        "kind": "INTERFACE",
        "name": "Node",
        "possibleTypes": [
          { "name": "Activity" },
          { "name": "Event" },
          { "name": "Community" },
          { "name": "User" },
          { "name": "Rsvp" },
          { "name": "EventInvitation" },
          { "name": "Message" },
          { "name": "Location" },
          { "name": "AppInvitation" },
          { "name": "AdminMessage" }
        ]
      },
      {
        "kind": "INTERFACE",
        "name": "Connection",
        "possibleTypes": [
          { "name": "ActivityEventsConnection" },
          { "name": "CommunityUsersConnection" },
          { "name": "UserRsvpsConnection" },
          { "name": "CommunityEventsConnection" },
          { "name": "EventRsvpsConnection" },
          { "name": "EventMessagesConnection" },
          { "name": "RootActivitiesConnection" },
          { "name": "RootCommunitiesConnection" },
          { "name": "RootEventsConnection" },
          { "name": "RootRsvpsConnection" },
          { "name": "RootLocationsConnection" },
          { "name": "AppInvitationConnection" },
          { "name": "RootUsersConnection" },
          { "name": "AdminMessagesConnection" }
        ]
      },
      {
        "kind": "INTERFACE",
        "name": "Edge",
        "possibleTypes": [
          { "name": "ActivityEventsEdge" },
          { "name": "CommunityUsersEdge" },
          { "name": "UserRsvpsEdge" },
          { "name": "CommunityEventsEdge" },
          { "name": "EventRsvpsEdge" },
          { "name": "EventMessagesEdge" },
          { "name": "RootActivitiesEdge" },
          { "name": "RootCommunitiesEdge" },
          { "name": "RootEventsEdge" },
          { "name": "RootRsvpsEdge" },
          { "name": "RootLocationsEdge" },
          { "name": "AppInvitationEdge" },
          { "name": "RootUsersEdge" },
          { "name": "AdminMessagesEdge" }
        ]
      },
      {
        "kind": "INTERFACE",
        "name": "Invitation",
        "possibleTypes": [
          { "name": "EventInvitation" },
          { "name": "AppInvitation" }
        ]
      },
      {
        "kind": "UNION",
        "name": "FieldValue",
        "possibleTypes": [
          { "name": "StringFieldValue" },
          { "name": "IntFieldValue" },
          { "name": "DateTimeFieldValue" }
        ]
      }
    ]
  }
}
