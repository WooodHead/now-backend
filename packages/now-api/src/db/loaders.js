import DataLoader from 'dataloader';
import { getUserBatch } from '../schema/resolvers/User';
import {
  Event,
  Activity,
  Rsvp,
  Location,
  Invitation,
  Community,
  Membership,
} from './repos';

const maxBatchSize = 100;
export default ({ currentUserId }) => ({
  members: new DataLoader(ids => getUserBatch(ids, currentUserId), {
    maxBatchSize,
  }),
  events: new DataLoader(ids => Event.batch(ids), {
    maxBatchSize,
  }),
  locations: new DataLoader(ids => Location.batch(ids), {
    maxBatchSize,
  }),
  activities: new DataLoader(ids => Activity.batch(ids), {
    maxBatchSize,
  }),
  rsvps: new DataLoader(ids => Rsvp.batch(ids), {
    maxBatchSize,
  }),
  invitations: new DataLoader(ids => Invitation.batch(ids), {
    maxBatchSize,
  }),
  communities: new DataLoader(ids => Community.batch(ids), {
    maxBatchSize,
  }),
  userMemberships: new DataLoader(
    ids => Promise.all(ids.map(id => Membership.all({ userId: id }))),
    {
      maxBatchSize,
    }
  ),
});
