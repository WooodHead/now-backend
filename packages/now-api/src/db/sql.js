import knex from 'knex';
import { setPgTypes } from './types';
import knexConfig from '../../knexfile';

const config = knexConfig[process.env.NODE_ENV];
const sql = knex(config);

setPgTypes(sql);

export default sql;
export const connection = config.connection; // eslint-disable-line prefer-destructuring

export const genRandomUuid = () => sql.raw('gen_random_uuid()');
export const makepoint = (x, y) => sql.raw('st_makepoint(?, ?)', x, y);
export const now = () => sql.raw('now()');
