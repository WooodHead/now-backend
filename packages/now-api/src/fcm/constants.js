/* eslint-disable import/prefer-default-export */
export const NEW_EVENTS_TOPIC = 'new-events';
