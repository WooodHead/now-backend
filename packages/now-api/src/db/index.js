import AWS from 'aws-sdk';
import promisify from 'util.promisify';
import { rsvpId } from '../schema/util';
import { TABLES } from './constants';
import { isDev } from '../util';

const dynamoOpts = {};
if (isDev()) {
  dynamoOpts.logger = console;
}

const dynamoDb = new AWS.DynamoDB.DocumentClient(dynamoOpts);

const pGet = promisify(dynamoDb.get.bind(dynamoDb));
const pScan = promisify(dynamoDb.scan.bind(dynamoDb));
const pUpdate = promisify(dynamoDb.update.bind(dynamoDb));
const pPut = promisify(dynamoDb.put.bind(dynamoDb));
const pQuery = promisify(dynamoDb.query.bind(dynamoDb));

export const get = (table, key = {}) =>
  pGet({
    TableName: table,
    Key: key,
  }).then(response => response.Item);

export const put = (table, item, condition = undefined) =>
  pPut({
    TableName: table,
    Item: item,
    ConditionExpression: condition,
  });

export const update = (
  table,
  key,
  expr,
  values,
  names = undefined,
  condition = undefined
) =>
  pUpdate({
    TableName: table,
    Key: key,
    ExpressionAttributeValues: values,
    ExpressionAttributeNames: names,
    UpdateExpression: expr,
    ConditionExpression: condition,
    ReturnValues: 'ALL_NEW',
  });

export const updateDynamic = (table, key, values, condition = undefined) => {
  const expressions = [];
  const expressionNames = {};
  const expressionValues = {};
  Object.entries(values).forEach(([k, v], i) => {
    const attrName = `#f${i}`;
    const fieldName = `:f${i}`;
    expressions.push(`${attrName}=${fieldName}`);
    expressionNames[attrName] = k;
    expressionValues[fieldName] = v;
  });

  return update(
    table,
    key,
    `set ${expressions.join(', ')}`,
    expressionValues,
    expressionNames,
    condition
  );
};

export const scan = (table, filter = undefined) =>
  pScan({
    TableName: table,
    FilterExpression: filter,
  }).then(response => response.Items);

export const queryRaw = params => pQuery(params);
export const query = params => pQuery(params).then(response => response.Items);

export const getUserRsvpByEvent = (userId, eventId) =>
  get(TABLES.RSVP, { id: rsvpId(eventId, userId) });

export const getActivity = id => get(TABLES.ACTIVITY, { id });

export const getEvent = id =>
  pQuery({
    TableName: TABLES.EVENT,
    KeyConditionExpression: 'id = :id',
    ExpressionAttributeValues: { ':id': id },
    IndexName: 'id-index',
  }).then(response => response.Items[0]);

export { TABLES };
