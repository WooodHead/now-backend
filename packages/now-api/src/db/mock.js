/* eslint-disable import/no-extraneous-dependencies  */
import ApolloClient from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';

import schema from '../schema';
import loaders from './loaders';

export const USER_ID = 'aea68a98-591c-11e8-9cdb-872a15ebfd30';
export const mockPromise = (resolver, rejecter) =>
  rejecter ? Promise.reject(rejecter) : Promise.resolve(resolver);

export const mocks = {
  loadMember: jest.fn(),
};

const apolloCache = new InMemoryCache();

let isAdmin = false;

export const setAdmin = b => {
  isAdmin = b;
};

export const client = new ApolloClient({
  cache: apolloCache,
  link: new SchemaLink({
    schema,
    context: () => ({
      token: null,
      user: { id: USER_ID },
      loaders: loaders({ currentUserId: USER_ID }),
      scopes: isAdmin ? ['admin'] : [],
    }),
  }),
});

export const newUserClient = (currentUserAuth0Id, currentUserId) =>
  new ApolloClient({
    cache: apolloCache,
    link: new SchemaLink({
      schema,
      context: {
        token: null,
        currentUserAuth0Id,
        loaders: loaders({ currentUserId }),
      },
    }),
  });
