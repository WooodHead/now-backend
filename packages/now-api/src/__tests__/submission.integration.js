import gql from 'graphql-tag';

import { client, USER_ID } from '../db/mock';
import sql from '../db/sql';
import { SQL_TABLES } from '../db/constants';
import factory from '../db/factory';

const truncateTables = () =>
  Promise.all([
    sql(SQL_TABLES.SUBMISSIONS).truncate(),
    sql(SQL_TABLES.USERS).truncate(),
  ]);

const user = factory.build('user', { id: USER_ID });

beforeEach(() =>
  truncateTables().then(() => sql(SQL_TABLES.USERS).insert(user)));

afterEach(() => truncateTables());

const submit = gql`
  mutation submit {
    submitTemplate(
      input: {
        templateId: "ebf997e5-cf24-407f-b98d-5108d73b4044"
        responses: [
          {
            fieldId: "42388918-defc-4364-b9dd-79f41019d511"
            dateTimeValue: "2020-10-18T18:30:00-04:00"
            durationValue: 6300000
          }
          {
            fieldId: "28e86c72-9961-4d0f-9fd1-9479ba148611"
            locationNameValue: "the museum"
            locationAddressValue: "One Museum Street"
          }
          {
            fieldId: "7071f2d9-85e8-430f-9f10-f94614cfb311"
            minValue: 4
            maxValue: 10
          }
          {
            fieldId: "3aaf4489-9ce1-409d-b656-a576ff4de011"
            stringValue: "we'll do some stuff"
          }
          {
            fieldId: "3df93f43-306d-44b5-9a39-d2d92f242011"
            stringValue: "it's for my mates"
          }
          {
            fieldId: "d6675c14-752d-4692-bcde-3894ff701011"
            stringValue: "knowledge is power"
          }
          {
            fieldId: "b5c15ff6-8e5e-4652-a3ac-8ca108fb1311"
            stringValue: "my awesome title"
          }
        ]
      }
    ) {
      submission {
        id
      }
    }
  }
`;

const query = gql`
  query submission($id: ID!) {
    submission(id: $id) {
      id
      template {
        id
        title
      }
      user {
        id
        firstName
        bio
      }
      createdAt
      updatedAt
      responses {
        id
        field {
          id
          type
        }
        value {
          __typename
          ... on StringFieldValue {
            s: value
          }
          ... on IntFieldValue {
            i: value
          }
          ... on DateTimeFieldValue {
            dt: value
            duration
          }
          ... on RangeFieldValue {
            min
            max
          }
          ... on LocationFieldValue {
            name
            address
          }
        }
      }
      formattedResponse
    }
  }
`;

describe('submission', () => {
  it('can post a submission and then retrieve it', async () => {
    const {
      data: {
        submitTemplate: {
          submission: { id },
        },
      },
    } = await client.mutate({ mutation: submit });

    const {
      data: { submission },
    } = await client.query({
      query,
      variables: { id },
    });

    expect(submission).toMatchObject({
      id: expect.stringMatching(/^[a-f0-9-]+$/),
      template: {
        id: 'ebf997e5-cf24-407f-b98d-5108d73b4044',
        title: 'Visit A Museum Exhibit',
      },
      user: {
        id: USER_ID,
        firstName: user.firstName,
        bio: user.bio,
      },
    });

    const { responses } = submission;
    expect(responses).toHaveLength(7);
    [
      'StringFieldValue',
      'DateTimeFieldValue',
      'RangeFieldValue',
      'LocationFieldValue',
    ].forEach(fieldType => {
      expect(responses).toContainEqual(
        expect.objectContaining({
          value: expect.objectContaining({ __typename: fieldType }),
        })
      );
    });
    responses.forEach(response => {
      expect(response).toHaveProperty(
        'field.type',
        expect.stringMatching(/^Text|NumberRange|Location|DateTimeDuration$/)
      );
      expect(response.id).toEqual(expect.any(String));
      // eslint-disable-next-line no-underscore-dangle
      switch (response.value.__typename) {
        case 'DateTimeFieldValue':
          expect(response.value.dt).toMatch(
            /^20[0-9]{2}-[0-9]{2}-[0-9]{2}T.*$/
          );
          expect(response.value.duration).toBe(6300000);
          break;

        case 'StringFieldValue':
          expect(response.value.s).toEqual(expect.any(String));
          break;

        case 'IntFieldValue':
          expect(response.value.i).toEqual(expect.any(Number));
          break;

        case 'RangeFieldValue':
          expect(response.value.min).toEqual(4);
          expect(response.value.max).toEqual(10);
          break;

        case 'LocationFieldValue':
          expect(response.value.name).toEqual('the museum');
          expect(response.value.address).toEqual('One Museum Street');
          break;

        default:
          expect(true).toBe(false); // shouldn't happen!
      }
    });
    expect(submission.formattedResponse).toMatchSnapshot();
  });
});
