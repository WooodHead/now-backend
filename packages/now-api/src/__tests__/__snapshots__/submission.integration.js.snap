// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`submission can post a submission and then retrieve it 1`] = `
"# my awesome title

## Details

at the museum  
One Museum Street  
2020\\\\-10\\\\-18T18:30\\\\-04:00  
105 minutes long  
4â€“10 spots

## Description

### Who's this for?

it's for my mates

### What will you do?

we'll do some stuff

### What should they know?

knowledge is power
"
`;
